sklearn/utils/isfinite.pyx

```diff
@@ -0,0 +1,152 @@
+
+    # conversion function
+    float npy_half_to_float(npy_half h);
+
+
+cdef fused fprecision:
+    npy_half
+    float
+    double
+    float complex
+    double complex
+
+cdef fused const_fprecision:
+    const npy_half
+    const float
+    const double
+    const float complex
+    const double complex
+
+
+cdef enum bint_enum:
+    false = False
+    true = True
+
+
+cdef struct bint_false_type:
+    char empty
+
+cdef struct bint_true_type:
+    char empty
+
+cdef fused bint_type:
+    bint_false_type
+    bint_true_type
+
+
+@cython.cdivision(True)
+def py_isfinite(numpy.ndarray a, bint allow_nan=False):
+    cdef numpy.ndarray a_flat
+
+    cdef char* a_data
+    cdef numpy.NPY_TYPES a_type
+    cdef size_t a_step
+    cdef size_t a_size
+
+    cdef bint disallow_nan
+
+    cdef bint result
+    cdef bint_enum err
+
+
+    a_flat = numpy.PyArray_Reshape(a, -1)
+
+    with nogil:
+        a_data = a_flat.data
+        a_type = <numpy.NPY_TYPES>a_flat.descr.type_num
+        a_step = a_flat.strides[0] / a_flat.descr.itemsize
+        a_size = a_flat.shape[0]
+
+        disallow_nan = not allow_nan
+
+        err = bint_enum.false
+
+        if a_type == numpy.NPY_BOOL:
+            result = True
+        elif a_type == numpy.NPY_BYTE:
+            result = True
+        elif a_type == numpy.NPY_UBYTE:
+            result = True
+        elif a_type == numpy.NPY_SHORT:
+            result = True
+        elif a_type == numpy.NPY_USHORT:
+            result = True
+        elif a_type == numpy.NPY_INT:
+            result = True
+        elif a_type == numpy.NPY_UINT:
+            result = True
+        elif a_type == numpy.NPY_LONG:
+            result = True
+        elif a_type == numpy.NPY_ULONG:
+            result = True
+        elif a_type == numpy.NPY_LONGLONG:
+            result = True
+        elif a_type == numpy.NPY_ULONGLONG:
+            result = True
+        elif a_type == numpy.NPY_HALF:
+            result = c_isfinite(<const npy_half*>a_data, a_step, a_size, <bint_enum>disallow_nan)
+        elif a_type == numpy.NPY_FLOAT:
+            result = c_isfinite(<const float*>a_data, a_step, a_size, <bint_enum>disallow_nan)
+        elif a_type == numpy.NPY_DOUBLE:
+            result = c_isfinite(<const double*>a_data, a_step, a_size, <bint_enum>disallow_nan)
+        elif a_type == numpy.NPY_CFLOAT:
+            result = c_isfinite(<const float complex*>a_data, a_step, a_size, <bint_enum>disallow_nan)
+        elif a_type == numpy.NPY_CDOUBLE:
+            result = c_isfinite(<const double complex*>a_data, a_step, a_size, <bint_enum>disallow_nan)
+        else:
+            err = bint_enum.true
+
+    if err == bint_enum.false:
+        return result
+    elif err == bint_enum.true:
+        raise TypeError("Unsupported array type: %s" % repr(numpy.PyArray_TypeObjectFromType(a_type)))
+
+
+cdef inline bint c_isfinite(const_fprecision* a_ptr, size_t step, size_t size, bint_enum disallow_nan) nogil:
```

Comment:
There is a lot of indirection for getting `disallow_nan` passed through. Although, I do not see the motivation for the `bint_enum`.

I suspect the original implementation is trying to template out the `bint_type` so the overhead of checking for `bint` is compiled away in the loop. I am interested to see what the overhead looks like without this optimization.

Reply:
I'll run some benchmarks once the rest of the pattern is stable -- I'm also interested to see the realized performance difference.

Reply:
For context the `bint_type` was intended to move a deeply nested branch from a repeated runtime check to a one time mostly compiled check. It borrows from C++ templating tricks. That said, there are other ways to achieve the same behavior.